File: CS/Grid MVVM Filtering/ViewModel/DataViewModel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Collections.ObjectModel;
using Grid_MVVM_Filtering.Model;
using System.Windows.Input;

namespace Grid_MVVM_Filtering.ViewModel
{
    class DataViewModel : ObservableObject
    {
        public DataViewModel(int length)
        {
            CreateData(length);
        }

        private void CreateData(int length)
        {
            Data = new ObservableCollection<DataModel>();
            VisibleData = new ObservableCollection<DataModel>();
            ExcludedData = new ObservableCollection<DataModel>();
            for (int i = 0; i < length; i++)
            {
                Data.Add(new DataModel { Text = "Row" + i, Number = i });
            }
        }

        ObservableCollection<DataModel> _Data;
        ObservableCollection<DataModel> _VisibleData;
        ObservableCollection<DataModel> _ExcludedData;
        RelayCommand<DataModel> _AddExclusionCommand;
        RelayCommand<DataModel> _RemoveExclusionCommand;

        public ObservableCollection<DataModel> Data
        {
            get
            {
                return _Data;
            }
            set
            {
                _Data = value;
                OnPropertyChanged("Data");
            }
        }

        public ObservableCollection<DataModel> VisibleData
        {
            get
            {
                return _VisibleData;
            }
            set
            {
                _VisibleData = value;
                OnPropertyChanged("VisibleData");
            }
        }

        public ObservableCollection<DataModel> ExcludedData
        {
            get
            {
                return _ExcludedData;
            }
            set
            {
                _ExcludedData = value;
                OnPropertyChanged("ExcludedData");
            }
        }

        public ICommand AddExclusionCommand
        {
            get
            {
                if (_AddExclusionCommand == null)
                {
                    _AddExclusionCommand = new RelayCommand<DataModel>(param => this.AddExclusion(param), param => this.CanAddExclusion(param));
                }
                return _AddExclusionCommand;
            }
        }

        private void AddExclusion(DataModel param)
        {
            ExcludedData.Add(param);
            OnPropertyChanged("ExcludedData");
        }

        private bool CanAddExclusion(DataModel param)
        {
            return param != null && Data.Contains(param) && !ExcludedData.Contains(param);
        }

        public ICommand RemoveExclusionCommand
        {
            get
            {
                if (_RemoveExclusionCommand == null)
                {
                    _RemoveExclusionCommand = new RelayCommand<DataModel>(param => this.RemoveExclusion(param), param => this.CanRemoveExclusion(param));
                }
                return _RemoveExclusionCommand;
            }
        }

        private void RemoveExclusion(DataModel param)
        {
            ExcludedData.Remove(param);
            OnPropertyChanged("ExcludedData");
        }

        private bool CanRemoveExclusion(DataModel param)
        {
            return param != null && Data.Contains(param) && ExcludedData.Contains(param);
        }
    }
}

File: VB/Grid MVVM Filtering/ViewModel/DataViewModel.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Collections.ObjectModel
Imports Grid_MVVM_Filtering.Model
Imports System.Windows.Input

Namespace Grid_MVVM_Filtering.ViewModel
	Friend Class DataViewModel
		Inherits ObservableObject
		Public Sub New(ByVal length As Integer)
			CreateData(length)
		End Sub

		Private Sub CreateData(ByVal length As Integer)
			Data = New ObservableCollection(Of DataModel)()
			VisibleData = New ObservableCollection(Of DataModel)()
			ExcludedData = New ObservableCollection(Of DataModel)()
			For i As Integer = 0 To length - 1
				Data.Add(New DataModel With {.Text = "Row" & i, .Number = i})
			Next i
		End Sub

		Private _Data As ObservableCollection(Of DataModel)
		Private _VisibleData As ObservableCollection(Of DataModel)
		Private _ExcludedData As ObservableCollection(Of DataModel)
		Private _AddExclusionCommand As RelayCommand(Of DataModel)
		Private _RemoveExclusionCommand As RelayCommand(Of DataModel)

		Public Property Data() As ObservableCollection(Of DataModel)
			Get
				Return _Data
			End Get
			Set(ByVal value As ObservableCollection(Of DataModel))
				_Data = value
				OnPropertyChanged("Data")
			End Set
		End Property

		Public Property VisibleData() As ObservableCollection(Of DataModel)
			Get
				Return _VisibleData
			End Get
			Set(ByVal value As ObservableCollection(Of DataModel))
				_VisibleData = value
				OnPropertyChanged("VisibleData")
			End Set
		End Property

		Public Property ExcludedData() As ObservableCollection(Of DataModel)
			Get
				Return _ExcludedData
			End Get
			Set(ByVal value As ObservableCollection(Of DataModel))
				_ExcludedData = value
				OnPropertyChanged("ExcludedData")
			End Set
		End Property

		Public ReadOnly Property AddExclusionCommand() As ICommand
			Get
				If _AddExclusionCommand Is Nothing Then
                    _AddExclusionCommand = New RelayCommand(Of DataModel)(Sub(param) Me.AddExclusion(param), Function(param) Me.CanAddExclusion(param))
				End If
				Return _AddExclusionCommand
			End Get
		End Property

		Private Sub AddExclusion(ByVal param As DataModel)
			ExcludedData.Add(param)
			OnPropertyChanged("ExcludedData")
		End Sub

		Private Function CanAddExclusion(ByVal param As DataModel) As Boolean
			Return param IsNot Nothing AndAlso Data.Contains(param) AndAlso Not ExcludedData.Contains(param)
		End Function

		Public ReadOnly Property RemoveExclusionCommand() As ICommand
			Get
				If _RemoveExclusionCommand Is Nothing Then
                    _RemoveExclusionCommand = New RelayCommand(Of DataModel)(Sub(param) Me.RemoveExclusion(param), Function(param) Me.CanRemoveExclusion(param))
				End If
				Return _RemoveExclusionCommand
			End Get
		End Property

		Private Sub RemoveExclusion(ByVal param As DataModel)
			ExcludedData.Remove(param)
			OnPropertyChanged("ExcludedData")
		End Sub

		Private Function CanRemoveExclusion(ByVal param As DataModel) As Boolean
			Return param IsNot Nothing AndAlso Data.Contains(param) AndAlso ExcludedData.Contains(param)
		End Function
	End Class
End Namespace
